File: test/language/array_types/cpp/PackingInterfaceOptimizationTest.cpp
#include "array_types/packing_interface_optimization/PackingInterfaceOptimization.h"
#include "gtest/gtest.h"
#include "test_utils/Assertions.h"
#include "zserio/RebindAlloc.h"
#include "zserio/SerializeUtil.h"

namespace array_types
{
namespace packing_interface_optimization
{

using allocator_type = PackingInterfaceOptimization::allocator_type;
template <typename T>
using vector_type = zserio::vector<T, allocator_type>;

class PackingInterfaceOptimizationTest : public ::testing::Test
{
protected:
    void fillUnpackedColors(vector_type<UnpackedColorStruct>& unpackedColors)
    {
        unpackedColors.emplace_back(true, UnpackedColorChoice{});
        unpackedColors.back().getColorChoice().setColorName("yellow");

        unpackedColors.emplace_back(false, UnpackedColorChoice{});
        unpackedColors.back().getColorChoice().setColorUnion(UnpackedColorUnion{});
        unpackedColors.back().getColorChoice().getColorUnion().setColorBitmask(
                UnpackedColorBitmask::Values::GREEN | UnpackedColorBitmask::Values::RED);

        unpackedColors.emplace_back(false, UnpackedColorChoice{});
        unpackedColors.back().getColorChoice().setColorUnion(UnpackedColorUnion{});
        unpackedColors.back().getColorChoice().getColorUnion().setColorEnum(UnpackedColorEnum::BLUE);
    }

    void fillMixedColors(vector_type<MixedColorStruct>& mixedColors)
    {
        mixedColors.emplace_back(true, MixedColorChoice{});
        mixedColors.back().getColorChoice().setColorName("purple");

        mixedColors.emplace_back(false, MixedColorChoice{});
        mixedColors.back().getColorChoice().setColorUnion(MixedColorUnion{});
        mixedColors.back().getColorChoice().getColorUnion().setColorBitmask(
                MixedColorBitmask::Values::BLUE | MixedColorBitmask::Values::GREEN);

        mixedColors.emplace_back(false, MixedColorChoice{});
        mixedColors.back().getColorChoice().setColorUnion(MixedColorUnion{});
        mixedColors.back().getColorChoice().getColorUnion().setColorEnum(MixedColorEnum::RED);
    }

    void fillPackedColors(vector_type<PackedColorStruct>& packedColors)
    {
        packedColors.emplace_back(true, PackedColorChoice{});
        packedColors.back().getColorChoice().setColorName("grey");

        packedColors.emplace_back(false, PackedColorChoice{});
        packedColors.back().getColorChoice().setColorUnion(PackedColorUnion{});
        packedColors.back().getColorChoice().getColorUnion().setColorBitmask(
                PackedColorBitmask::Values::BLUE | PackedColorBitmask::Values::RED);

        packedColors.emplace_back(false, PackedColorChoice{});
        packedColors.back().getColorChoice().setColorUnion(PackedColorUnion{});
        packedColors.back().getColorChoice().getColorUnion().setColorEnum(PackedColorEnum::GREEN);
    }

    void fillUnpackedColorsHolder(UnpackedColorsHolder& unpackedColorsHolder)
    {
        fillUnpackedColors(unpackedColorsHolder.getUnpackedColors());
        fillMixedColors(unpackedColorsHolder.getMixedColors());
    }

    void fillPackedColorsHolder(PackedColorsHolder& packedColorsHolder)
    {
        fillMixedColors(packedColorsHolder.getMixedColors());
        fillPackedColors(packedColorsHolder.getPackedColors());
    }

    void assertPackingInterfaceMethodsPresent(const std::string& typeName)
    {
        const std::string ctorDeclaration = typeName + "(ZserioPackingContext";
        const std::string ctorDefinition =
                typeName + "::" + typeName + "(" + typeName + "::ZserioPackingContext";
        ASSERT_METHOD_PRESENT(PATH, typeName, ctorDeclaration.c_str(), ctorDefinition.c_str());

        ASSERT_METHOD_PRESENT(PATH, typeName, "initPackingContext(", "::initPackingContext(");

        const std::string bitSizeOfMethodDeclaration = "bitSizeOf(ZserioPackingContext";
        const std::string bitSizeOfMethodDefinition = "bitSizeOf(" + typeName + "::ZserioPackingContext";
        ASSERT_METHOD_PRESENT(
                PATH, typeName, bitSizeOfMethodDeclaration.c_str(), bitSizeOfMethodDefinition.c_str());

        const std::string initializeOffsetsMethodDeclaration = "initializeOffsets(ZserioPackingContext";
        const std::string initializeOffsetsMethodDefinition =
                "initializeOffsets(" + typeName + "::ZserioPackingContext";
        ASSERT_METHOD_PRESENT(PATH, typeName, initializeOffsetsMethodDeclaration.c_str(),
                initializeOffsetsMethodDefinition.c_str());

        const std::string writeMethodDeclaration = "write(ZserioPackingContext";
        const std::string writeMethodDefinition = "write(" + typeName + "::ZserioPackingContext";
        ASSERT_METHOD_PRESENT(PATH, typeName, writeMethodDeclaration.c_str(), writeMethodDefinition.c_str());
    }

    void assertPackingInterfaceMethodsNotPresent(const std::string& typeName)
    {
        const std::string ctorDeclaration = typeName + "(ZserioPackingContext";
        const std::string ctorDefinition =
                typeName + "::" + typeName + "(" + typeName + "::ZserioPackingContext";
        ASSERT_METHOD_NOT_PRESENT(PATH, typeName, ctorDeclaration.c_str(), ctorDefinition.c_str());

        ASSERT_METHOD_NOT_PRESENT(PATH, typeName, "initPackingContext(", "::initPackingContext(");

        const std::string bitSizeOfMethodDeclaration = "bitSizeOf(ZserioPackingContext";
        const std::string bitSizeOfMethodDefinition = "bitSizeOf(" + typeName + "::ZserioPackingContext";
        ASSERT_METHOD_NOT_PRESENT(
                PATH, typeName, bitSizeOfMethodDeclaration.c_str(), bitSizeOfMethodDefinition.c_str());

        const std::string initializeOffsetsMethodDeclaration = "initializeOffsets(ZserioPackingContext";
        const std::string initializeOffsetsMethodDefinition =
                "initializeOffsets(" + typeName + "::ZserioPackingContext";
        ASSERT_METHOD_NOT_PRESENT(PATH, typeName, initializeOffsetsMethodDeclaration.c_str(),
                initializeOffsetsMethodDefinition.c_str());

        const std::string writeMethodDeclaration = "write(ZserioPackingContext";
        const std::string writeMethodDefinition = "write(" + typeName + "::ZserioPackingContext";
        ASSERT_METHOD_NOT_PRESENT(
                PATH, typeName, writeMethodDeclaration.c_str(), writeMethodDefinition.c_str());
    }

    static const std::string BLOB_NAME;
    static const std::string PATH;
};

const std::string PackingInterfaceOptimizationTest::BLOB_NAME =
        "language/array_types/packing_interface_optimization.blob";
const std::string PackingInterfaceOptimizationTest::PATH =
        "language/array_types/gen/array_types/packing_interface_optimization/";

TEST_F(PackingInterfaceOptimizationTest, writeReadFile)
{
    PackingInterfaceOptimization packingInterfaceOptimization;
    fillUnpackedColorsHolder(packingInterfaceOptimization.getUnpackedColorsHolder());
    fillPackedColorsHolder(packingInterfaceOptimization.getPackedColorsHolder());

    zserio::serializeToFile(packingInterfaceOptimization, BLOB_NAME);
    const auto readPackingInterfaceOptimization =
            zserio::deserializeFromFile<PackingInterfaceOptimization>(BLOB_NAME);
    ASSERT_EQ(packingInterfaceOptimization, readPackingInterfaceOptimization);
}

TEST_F(PackingInterfaceOptimizationTest, packingInterfaceOptimizationMethods)
{
    assertPackingInterfaceMethodsNotPresent("PackingInterfaceOptimization");
    ASSERT_STRING_IN_FILE_NOT_PRESENT(PATH + "PackingInterfaceOptimization.h", "class ZserioPackingContext");
}

TEST_F(PackingInterfaceOptimizationTest, unpackedColorsHolderMethods)
{
    assertPackingInterfaceMethodsNotPresent("UnpackedColorsHolder");
    ASSERT_STRING_IN_FILE_NOT_PRESENT(PATH + "UnpackedColorsHolder.h", "class ZserioPackingContext");

    ASSERT_STRING_IN_FILE_NOT_PRESENT(PATH + "UnpackedColorsHolder.h",
            "::array_types::packing_interface_optimization::UnpackedColorStruct::ZserioPackingContext& "
            "context");

    ASSERT_STRING_IN_FILE_PRESENT(PATH + "UnpackedColorsHolder.h",
            "::array_types::packing_interface_optimization::MixedColorStruct::ZserioPackingContext& context");
}

TEST_F(PackingInterfaceOptimizationTest, packedColorsHolderMethods)
{
    assertPackingInterfaceMethodsNotPresent("PackedColorsHolder");
    ASSERT_STRING_IN_FILE_NOT_PRESENT(PATH + "PackedColorsHolder.h", "class ZserioPackingContext");

    ASSERT_STRING_IN_FILE_PRESENT(PATH + "PackedColorsHolder.h",
            "::array_types::packing_interface_optimization::MixedColorStruct::ZserioPackingContext& context");

    ASSERT_STRING_IN_FILE_PRESENT(PATH + "PackedColorsHolder.h",
            "::array_types::packing_interface_optimization::PackedColorStruct::ZserioPackingContext& context");
}

TEST_F(PackingInterfaceOptimizationTest, unpackedColorStructMethods)
{
    assertPackingInterfaceMethodsNotPresent("UnpackedColorStruct");
    ASSERT_STRING_IN_FILE_NOT_PRESENT(PATH + "UnpackedColorStruct.h", "class ZserioPackingContext");
}

TEST_F(PackingInterfaceOptimizationTest, unpackedColorChoiceMethods)
{
    assertPackingInterfaceMethodsNotPresent("UnpackedColorChoice");
    ASSERT_STRING_IN_FILE_NOT_PRESENT(PATH + "UnpackedColorChoice.h", "class ZserioPackingContext");
}

TEST_F(PackingInterfaceOptimizationTest, unpackedColorUnionMethods)
{
    assertPackingInterfaceMethodsNotPresent("UnpackedColorUnion");
    ASSERT_STRING_IN_FILE_NOT_PRESENT(PATH + "UnpackedColorUnion.h", "class ZserioPackingContext");
}

TEST_F(PackingInterfaceOptimizationTest, unpackedColorEnumMethods)
{
    ASSERT_METHOD_NOT_PRESENT(PATH, "UnpackedColorEnum", "initPackingContext<::zserio::DeltaContext,",
            "initPackingContext(::zserio::DeltaContext&");
    ASSERT_METHOD_NOT_PRESENT(PATH, "UnpackedColorEnum", "bitSizeOf<::zserio::DeltaContext,",
            "bitSizeOf(::zserio::DeltaContext&");
    ASSERT_METHOD_NOT_PRESENT(PATH, "UnpackedColorEnum", "initializeOffsets<::zserio::DeltaContext,",
            "initializeOffsets(::zserio::DeltaContext&");
    ASSERT_METHOD_NOT_PRESENT(PATH, "UnpackedColorEnum",
            "read<::array_types::packing_interface_optimization::UnpackedColorEnum, ::zserio::DeltaContext",
            "read(::zserio::DeltaContext&");
    ASSERT_METHOD_NOT_PRESENT(
            PATH, "UnpackedColorEnum", "write<::zserio::DeltaContext,", "write(::zserio::DeltaContext&");
}

TEST_F(PackingInterfaceOptimizationTest, unpackedColorBitmaskMethods)
{
    ASSERT_METHOD_NOT_PRESENT(PATH, "UnpackedColorBitmask", "UnpackedColorBitmask(::zserio::DeltaContext",
            "UnpackedColorBitmask::UnpackedColorBitmask(::zserio::DeltaContext");

    ASSERT_METHOD_NOT_PRESENT(PATH, "UnpackedColorBitmask", "initPackingContext(", "::initPackingContext(");
    ASSERT_METHOD_NOT_PRESENT(PATH, "UnpackedColorBitmask", "bitSizeOf(::zserio::DeltaContext",
            "::bitSizeOf(::zserio::DeltaContext");
    ASSERT_METHOD_NOT_PRESENT(PATH, "UnpackedColorBitmask", "initializeOffsets(::zserio::DeltaContext",
            "::initializeOffsets(::zserio::DeltaContext");
    ASSERT_METHOD_NOT_PRESENT(
            PATH, "UnpackedColorBitmask", "write(::zserio::DeltaContext", "::write(::zserio::DeltaContext");
}

TEST_F(PackingInterfaceOptimizationTest, mixedColorStructMethods)
{
    assertPackingInterfaceMethodsPresent("MixedColorStruct");
    ASSERT_STRING_IN_FILE_PRESENT(PATH + "MixedColorStruct.h", "class ZserioPackingContext");
}

TEST_F(PackingInterfaceOptimizationTest, mixedColorChoiceMethods)
{
    assertPackingInterfaceMethodsPresent("MixedColorChoice");
    ASSERT_STRING_IN_FILE_PRESENT(PATH + "MixedColorChoice.h", "class ZserioPackingContext");
}

TEST_F(PackingInterfaceOptimizationTest, mixedColorUnionMethods)
{
    assertPackingInterfaceMethodsPresent("MixedColorUnion");
    ASSERT_STRING_IN_FILE_PRESENT(PATH + "MixedColorUnion.h", "class ZserioPackingContext");
}

TEST_F(PackingInterfaceOptimizationTest, mixedColorEnumMethods)
{
    ASSERT_METHOD_PRESENT(PATH, "MixedColorEnum", "initPackingContext<::zserio::DeltaContext,",
            "initPackingContext(::zserio::DeltaContext&");
    ASSERT_METHOD_PRESENT(
            PATH, "MixedColorEnum", "bitSizeOf<::zserio::DeltaContext,", "bitSizeOf(::zserio::DeltaContext&");
    ASSERT_METHOD_PRESENT(PATH, "MixedColorEnum", "initializeOffsets<::zserio::DeltaContext,",
            "initializeOffsets(::zserio::DeltaContext&");
    ASSERT_METHOD_PRESENT(PATH, "MixedColorEnum",
            "read<::array_types::packing_interface_optimization::MixedColorEnum, ::zserio::DeltaContext",
            "read(::zserio::DeltaContext&");
    ASSERT_METHOD_PRESENT(
            PATH, "MixedColorEnum", "write<::zserio::DeltaContext,", "write(::zserio::DeltaContext&");
}

TEST_F(PackingInterfaceOptimizationTest, mixedColorBitmaskMethods)
{
    ASSERT_METHOD_PRESENT(PATH, "MixedColorBitmask", "MixedColorBitmask(::zserio::DeltaContext",
            "MixedColorBitmask::MixedColorBitmask(::zserio::DeltaContext");

    ASSERT_METHOD_PRESENT(PATH, "MixedColorBitmask", "initPackingContext(", "::initPackingContext(");
    ASSERT_METHOD_PRESENT(PATH, "MixedColorBitmask", "bitSizeOf(::zserio::DeltaContext",
            "::bitSizeOf(::zserio::DeltaContext");
    ASSERT_METHOD_PRESENT(PATH, "MixedColorBitmask", "initializeOffsets(::zserio::DeltaContext",
            "::initializeOffsets(::zserio::DeltaContext");
    ASSERT_METHOD_PRESENT(
            PATH, "MixedColorBitmask", "write(::zserio::DeltaContext", "::write(::zserio::DeltaContext");
}

TEST_F(PackingInterfaceOptimizationTest, packedColorStructMethods)
{
    assertPackingInterfaceMethodsPresent("PackedColorStruct");
    ASSERT_STRING_IN_FILE_PRESENT(PATH + "PackedColorStruct.h", "class ZserioPackingContext");
}

TEST_F(PackingInterfaceOptimizationTest, packedColorChoiceMethods)
{
    assertPackingInterfaceMethodsPresent("PackedColorChoice");
    ASSERT_STRING_IN_FILE_PRESENT(PATH + "PackedColorChoice.h", "class ZserioPackingContext");
}

TEST_F(PackingInterfaceOptimizationTest, packedColorUnionMethods)
{
    assertPackingInterfaceMethodsPresent("PackedColorUnion");
    ASSERT_STRING_IN_FILE_PRESENT(PATH + "PackedColorUnion.h", "class ZserioPackingContext");
}

TEST_F(PackingInterfaceOptimizationTest, packedColorEnumMethods)
{
    ASSERT_METHOD_PRESENT(PATH, "PackedColorEnum", "initPackingContext<::zserio::DeltaContext,",
            "initPackingContext(::zserio::DeltaContext&");
    ASSERT_METHOD_PRESENT(
            PATH, "PackedColorEnum", "bitSizeOf<::zserio::DeltaContext,", "bitSizeOf(::zserio::DeltaContext&");
    ASSERT_METHOD_PRESENT(PATH, "PackedColorEnum", "initializeOffsets<::zserio::DeltaContext,",
            "initializeOffsets(::zserio::DeltaContext&");
    ASSERT_METHOD_PRESENT(PATH, "PackedColorEnum",
            "read<::array_types::packing_interface_optimization::PackedColorEnum, ::zserio::DeltaContext",
            "read(::zserio::DeltaContext&");
    ASSERT_METHOD_PRESENT(
            PATH, "PackedColorEnum", "write<::zserio::DeltaContext,", "write(::zserio::DeltaContext&");
}

TEST_F(PackingInterfaceOptimizationTest, packedColorBitmaskMethods)
{
    ASSERT_METHOD_PRESENT(PATH, "PackedColorBitmask", "PackedColorBitmask(::zserio::DeltaContext",
            "PackedColorBitmask::PackedColorBitmask(::zserio::DeltaContext");

    ASSERT_METHOD_PRESENT(PATH, "PackedColorBitmask", "initPackingContext(", "::initPackingContext(");
    ASSERT_METHOD_PRESENT(PATH, "PackedColorBitmask", "bitSizeOf(::zserio::DeltaContext",
            "::bitSizeOf(::zserio::DeltaContext");
    ASSERT_METHOD_PRESENT(PATH, "PackedColorBitmask", "initializeOffsets(::zserio::DeltaContext",
            "::initializeOffsets(::zserio::DeltaContext");
    ASSERT_METHOD_PRESENT(
            PATH, "PackedColorBitmask", "write(::zserio::DeltaContext", "::write(::zserio::DeltaContext");
}

} // namespace packing_interface_optimization
} // namespace array_types

File: test/extensions/language/array_types/java/array_types/PackingInterfaceOptimizationTest.java
package array_types;

import static org.junit.jupiter.api.Assertions.*;

import static test_utils.AssertionUtils.assertInnerClassNotPresent;
import static test_utils.AssertionUtils.assertInnerClassPresent;
import static test_utils.AssertionUtils.assertMethodNotPresent;
import static test_utils.AssertionUtils.assertMethodPresent;

import org.junit.jupiter.api.Test;

import zserio.runtime.io.SerializeUtil;

import array_types.packing_interface_optimization.MixedColorBitmask;
import array_types.packing_interface_optimization.MixedColorChoice;
import array_types.packing_interface_optimization.MixedColorEnum;
import array_types.packing_interface_optimization.MixedColorStruct;
import array_types.packing_interface_optimization.MixedColorUnion;
import array_types.packing_interface_optimization.PackedColorBitmask;
import array_types.packing_interface_optimization.PackedColorChoice;
import array_types.packing_interface_optimization.PackedColorEnum;
import array_types.packing_interface_optimization.PackedColorStruct;
import array_types.packing_interface_optimization.PackedColorUnion;
import array_types.packing_interface_optimization.PackedColorsHolder;
import array_types.packing_interface_optimization.PackingInterfaceOptimization;
import array_types.packing_interface_optimization.UnpackedColorBitmask;
import array_types.packing_interface_optimization.UnpackedColorChoice;
import array_types.packing_interface_optimization.UnpackedColorEnum;
import array_types.packing_interface_optimization.UnpackedColorStruct;
import array_types.packing_interface_optimization.UnpackedColorUnion;
import array_types.packing_interface_optimization.UnpackedColorsHolder;

public class PackingInterfaceOptimizationTest
{
    @Test
    public void writeReadFile()
    {
        final PackingInterfaceOptimization packingInterfaceOptimization =
                new PackingInterfaceOptimization(createUnpackedColorsHolder(), createPackedColorsHolder());

        SerializeUtil.serializeToFile(packingInterfaceOptimization, BLOB_NAME);
        final PackingInterfaceOptimization readPackingInterfaceOptimization =
                SerializeUtil.deserializeFromFile(PackingInterfaceOptimization.class, BLOB_NAME);
        assertEquals(packingInterfaceOptimization, readPackingInterfaceOptimization);
    }

    @Test
    public void packingInterfaceMethods()
    {
        assertPackingInterfaceMethodsNotPresent(PackingInterfaceOptimization.class);
        assertInnerClassNotPresent(PackingInterfaceOptimization.class, "ZserioPackingContext");
    }

    @Test
    public void unpackedColorsHolderMethods()
    {
        assertPackingInterfaceMethodsNotPresent(UnpackedColorsHolder.class);
        assertInnerClassNotPresent(UnpackedColorsHolder.class, "ZserioPackingContext");
        assertInnerClassPresent(UnpackedColorsHolder.class, "ZserioElementFactory_unpackedColors");
        assertInnerClassPresent(UnpackedColorsHolder.class, "ZserioElementFactory_mixedColors");

        final Class<?> zserioElementFactory_unpackedColors_class =
                getPrivateInnerClass(UnpackedColorsHolder.class, "ZserioElementFactory_unpackedColors");
        assertMethodPresent(zserioElementFactory_unpackedColors_class, "create");
        assertMethodNotPresent(zserioElementFactory_unpackedColors_class, "createPackingContext");
        assertMethodNotPresent(
                zserioElementFactory_unpackedColors_class, "create(zserio.runtime.array.PackingContext");

        final Class<?> zserioElementFactory_mixedColors_class =
                getPrivateInnerClass(UnpackedColorsHolder.class, "ZserioElementFactory_mixedColors");
        assertMethodPresent(zserioElementFactory_mixedColors_class, "create");
        assertMethodPresent(zserioElementFactory_mixedColors_class, "createPackingContext");
        assertMethodPresent(
                zserioElementFactory_mixedColors_class, "create(zserio.runtime.array.PackingContext");
    }

    @Test
    public void packedColorsHolderMethods()
    {
        assertPackingInterfaceMethodsNotPresent(PackedColorsHolder.class);
        assertInnerClassNotPresent(PackedColorsHolder.class, "ZserioPackingContext");
        assertInnerClassPresent(PackedColorsHolder.class, "ZserioElementFactory_mixedColors");
        assertInnerClassPresent(PackedColorsHolder.class, "ZserioElementFactory_packedColors");

        final Class<?> zserioElementFactory_mixedColors_class =
                getPrivateInnerClass(PackedColorsHolder.class, "ZserioElementFactory_mixedColors");
        assertMethodPresent(zserioElementFactory_mixedColors_class, "create");
        assertMethodPresent(zserioElementFactory_mixedColors_class, "createPackingContext");
        assertMethodPresent(
                zserioElementFactory_mixedColors_class, "create(zserio.runtime.array.PackingContext");

        final Class<?> zserioElementFactory_packedColors_class =
                getPrivateInnerClass(PackedColorsHolder.class, "ZserioElementFactory_packedColors");
        assertMethodPresent(zserioElementFactory_packedColors_class, "create");
        assertMethodPresent(zserioElementFactory_packedColors_class, "createPackingContext");
        assertMethodPresent(
                zserioElementFactory_packedColors_class, "create(zserio.runtime.array.PackingContext");
    }

    @Test
    public void unpackedColorStructMethods()
    {
        assertPackingInterfaceMethodsNotPresent(UnpackedColorStruct.class);
        assertInnerClassNotPresent(UnpackedColorStruct.class, "ZserioPackingContext");
    }

    @Test
    public void unpackedColorChoiceMethods()
    {
        assertPackingInterfaceMethodsNotPresent(UnpackedColorChoice.class);
        assertInnerClassNotPresent(UnpackedColorChoice.class, "ZserioPackingContext");
    }

    @Test
    public void unpackedColorUnionMethods()
    {
        assertPackingInterfaceMethodsNotPresent(UnpackedColorUnion.class);
        assertInnerClassNotPresent(UnpackedColorUnion.class, "ZserioPackingContext");
    }

    @Test
    public void unpackedColorEnumMethods()
    {
        assertMethodNotPresent(UnpackedColorEnum.class, "readEnum(zserio.runtime.array.PackingContext");
        assertMethodNotPresent(UnpackedColorEnum.class, "initPackingContext");
        assertMethodNotPresent(UnpackedColorEnum.class, "bitSizeOf(zserio.runtime.array.PackingContext");
        assertMethodNotPresent(
                UnpackedColorEnum.class, "initializeOffsets(zserio.runtime.array.PackingContext");
        assertMethodNotPresent(UnpackedColorEnum.class, "write(zserio.runtime.array.PackingContext");
    }

    @Test
    public void unpackedColorBitmaskMethods()
    {
        assertMethodNotPresent(
                UnpackedColorBitmask.class, "UnpackedColorBitmask(zserio.runtime.array.PackingContext");
        assertMethodNotPresent(UnpackedColorBitmask.class, "initPackingContext");
        assertMethodNotPresent(UnpackedColorBitmask.class, "bitSizeOf(zserio.runtime.array.PackingContext");
        assertMethodNotPresent(
                UnpackedColorBitmask.class, "initializeOffsets(zserio.runtime.array.PackingContext");
        assertMethodNotPresent(UnpackedColorBitmask.class, "write(zserio.runtime.array.PackingContext");
    }

    @Test
    public void mixedColorStructMethods()
    {
        assertPackingInterfaceMethodsPresent(MixedColorStruct.class);
        assertInnerClassPresent(MixedColorStruct.class, "ZserioPackingContext");
    }

    @Test
    public void mixedColorChoiceMethods()
    {
        assertPackingInterfaceMethodsPresent(MixedColorChoice.class);
        assertInnerClassPresent(MixedColorChoice.class, "ZserioPackingContext");
    }

    @Test
    public void mixedColorUnionMethods()
    {
        assertPackingInterfaceMethodsPresent(MixedColorUnion.class);
        assertInnerClassPresent(MixedColorUnion.class, "ZserioPackingContext");
    }

    @Test
    public void mixedColorEnumMethods()
    {
        assertMethodPresent(MixedColorEnum.class, "readEnum(zserio.runtime.array.PackingContext");
        assertMethodPresent(MixedColorEnum.class, "initPackingContext");
        assertMethodPresent(MixedColorEnum.class, "bitSizeOf(zserio.runtime.array.PackingContext");
        assertMethodPresent(MixedColorEnum.class, "initializeOffsets(zserio.runtime.array.PackingContext");
        assertMethodPresent(MixedColorEnum.class, "write(zserio.runtime.array.PackingContext");
    }

    @Test
    public void mixedColorBitmaskMethods()
    {
        assertMethodPresent(MixedColorBitmask.class, "MixedColorBitmask(zserio.runtime.array.PackingContext");
        assertMethodPresent(MixedColorBitmask.class, "initPackingContext");
        assertMethodPresent(MixedColorBitmask.class, "bitSizeOf(zserio.runtime.array.PackingContext");
        assertMethodPresent(MixedColorBitmask.class, "initializeOffsets(zserio.runtime.array.PackingContext");
        assertMethodPresent(MixedColorBitmask.class, "write(zserio.runtime.array.PackingContext");
    }

    @Test
    public void packedColorStructMethods()
    {
        assertPackingInterfaceMethodsPresent(PackedColorStruct.class);
        assertInnerClassPresent(PackedColorStruct.class, "ZserioPackingContext");
    }

    @Test
    public void packedColorChoiceMethods()
    {
        assertPackingInterfaceMethodsPresent(PackedColorChoice.class);
        assertInnerClassPresent(PackedColorChoice.class, "ZserioPackingContext");
    }

    @Test
    public void packedColorUnionMethods()
    {
        assertPackingInterfaceMethodsPresent(PackedColorUnion.class);
        assertInnerClassPresent(PackedColorUnion.class, "ZserioPackingContext");
    }

    @Test
    public void packedColorEnumMethods()
    {
        assertMethodPresent(PackedColorEnum.class, "readEnum(zserio.runtime.array.PackingContext");
        assertMethodPresent(PackedColorEnum.class, "initPackingContext");
        assertMethodPresent(PackedColorEnum.class, "bitSizeOf(zserio.runtime.array.PackingContext");
        assertMethodPresent(PackedColorEnum.class, "initializeOffsets(zserio.runtime.array.PackingContext");
        assertMethodPresent(PackedColorEnum.class, "write(zserio.runtime.array.PackingContext");
    }

    @Test
    public void packedColorBitmaskMethods()
    {
        assertMethodPresent(PackedColorBitmask.class, "PackedColorBitmask(zserio.runtime.array.PackingContext");
        assertMethodPresent(PackedColorBitmask.class, "initPackingContext");
        assertMethodPresent(PackedColorBitmask.class, "bitSizeOf(zserio.runtime.array.PackingContext");
        assertMethodPresent(PackedColorBitmask.class, "initializeOffsets(zserio.runtime.array.PackingContext");
        assertMethodPresent(PackedColorBitmask.class, "write(zserio.runtime.array.PackingContext");
    }

    private void assertPackingInterfaceMethodsNotPresent(Class<?> userType)
    {
        assertMethodNotPresent(userType, userType.getSimpleName() + "(zserio.runtime.array.PackingContext");
        assertMethodNotPresent(userType, "initPackingContext");
        assertMethodNotPresent(userType, "bitSizeOf(zserio.runtime.array.PackingContext");
        assertMethodNotPresent(userType, "initializeOffsets(zserio.runtime.array.PackingContext");
        assertMethodNotPresent(userType, "read(zserio.runtime.array.PackingContext");
        assertMethodNotPresent(userType, "write(zserio.runtime.array.PackingContext");
    }

    private void assertPackingInterfaceMethodsPresent(Class<?> userType)
    {
        assertMethodPresent(userType, userType.getSimpleName() + "(zserio.runtime.array.PackingContext");
        assertMethodPresent(userType, "initPackingContext");
        assertMethodPresent(userType, "bitSizeOf(zserio.runtime.array.PackingContext");
        assertMethodPresent(userType, "initializeOffsets(zserio.runtime.array.PackingContext");
        assertMethodPresent(userType, "read(zserio.runtime.array.PackingContext");
        assertMethodPresent(userType, "write(zserio.runtime.array.PackingContext");
    }

    private Class<?> getPrivateInnerClass(Class<?> userType, String innerClassName)
    {
        for (Class<?> innerClass : userType.getDeclaredClasses())
        {
            if (innerClass.getSimpleName().equals(innerClassName))
                return innerClass;
        }

        return null;
    }

    private UnpackedColorStruct[] createUnpackedColors()
    {
        final UnpackedColorChoice unpackedColorChoice1 = new UnpackedColorChoice(true);
        unpackedColorChoice1.setColorName("yellow");
        final UnpackedColorChoice unpackedColorChoice2 = new UnpackedColorChoice(false);
        final UnpackedColorUnion unpackedColorUnion2 = new UnpackedColorUnion();
        unpackedColorUnion2.setColorBitmask(
                UnpackedColorBitmask.Values.GREEN.or(UnpackedColorBitmask.Values.RED));
        unpackedColorChoice2.setColorUnion(unpackedColorUnion2);
        final UnpackedColorChoice unpackedColorChoice3 = new UnpackedColorChoice(false);
        final UnpackedColorUnion unpackedColorUnion3 = new UnpackedColorUnion();
        unpackedColorUnion3.setColorEnum(UnpackedColorEnum.BLUE);
        unpackedColorChoice3.setColorUnion(unpackedColorUnion3);
        return new UnpackedColorStruct[] {
                new UnpackedColorStruct(unpackedColorChoice1.getSelector(), unpackedColorChoice1),
                new UnpackedColorStruct(unpackedColorChoice2.getSelector(), unpackedColorChoice2),
                new UnpackedColorStruct(unpackedColorChoice3.getSelector(), unpackedColorChoice3)};
    }

    private MixedColorStruct[] createMixedColors()
    {
        final MixedColorChoice mixedColorChoice1 = new MixedColorChoice(true);
        mixedColorChoice1.setColorName("purple");
        final MixedColorChoice mixedColorChoice2 = new MixedColorChoice(false);
        final MixedColorUnion mixedColorUnion2 = new MixedColorUnion();
        mixedColorUnion2.setColorBitmask(MixedColorBitmask.Values.BLUE.or(MixedColorBitmask.Values.GREEN));
        mixedColorChoice2.setColorUnion(mixedColorUnion2);
        final MixedColorChoice mixedColorChoice3 = new MixedColorChoice(false);
        final MixedColorUnion mixedColorUnion3 = new MixedColorUnion();
        mixedColorUnion3.setColorEnum(MixedColorEnum.RED);
        mixedColorChoice3.setColorUnion(mixedColorUnion3);
        return new MixedColorStruct[] {new MixedColorStruct(mixedColorChoice1.getSelector(), mixedColorChoice1),
                new MixedColorStruct(mixedColorChoice2.getSelector(), mixedColorChoice2),
                new MixedColorStruct(mixedColorChoice3.getSelector(), mixedColorChoice3)};
    }

    private PackedColorStruct[] createPackedColors()
    {
        final PackedColorChoice packedColorChoice1 = new PackedColorChoice(true);
        packedColorChoice1.setColorName("grey");
        final PackedColorChoice packedColorChoice2 = new PackedColorChoice(false);
        final PackedColorUnion packedColorUnion2 = new PackedColorUnion();
        packedColorUnion2.setColorBitmask(PackedColorBitmask.Values.BLUE.or(PackedColorBitmask.Values.RED));
        packedColorChoice2.setColorUnion(packedColorUnion2);
        final PackedColorChoice packedColorChoice3 = new PackedColorChoice(false);
        final PackedColorUnion packedColorUnion3 = new PackedColorUnion();
        packedColorUnion3.setColorEnum(PackedColorEnum.GREEN);
        packedColorChoice3.setColorUnion(packedColorUnion3);
        return new PackedColorStruct[] {
                new PackedColorStruct(packedColorChoice1.getSelector(), packedColorChoice1),
                new PackedColorStruct(packedColorChoice2.getSelector(), packedColorChoice2),
                new PackedColorStruct(packedColorChoice3.getSelector(), packedColorChoice3)};
    }

    private UnpackedColorsHolder createUnpackedColorsHolder()
    {
        return new UnpackedColorsHolder(createUnpackedColors(), createMixedColors());
    }

    private PackedColorsHolder createPackedColorsHolder()
    {
        return new PackedColorsHolder(createMixedColors(), createPackedColors());
    }

    private static final String BLOB_NAME = "packing_interface_optimization.blob";
}
